cdc/sink/dmlsink/mq/mq_dml_sink.go

```diff
@@ -144,10 +147,16 @@ func (s *dmlSink) WriteEvents(txns ...*dmlsink.CallbackableEvent[*model.SingleTa
 			continue
 		}
 		callback := mergedCallback(txn.Callback, uint64(len(txn.Event.Rows)))
+
 		for _, row := range txn.Event.Rows {
 			topic := s.alive.eventRouter.GetTopicForRowChange(row)
 			partitionNum, err := s.alive.topicManager.GetPartitionNum(s.ctx, topic)
+			failpoint.Inject("MQSinkGetPartitionError", func() {
+				log.Info("failpoint MQSinkGetPartitionError injected", zap.String("changefeedID", s.id.ID))
+				err = errors.New("MQSinkGetPartitionError")
+			})
 			if err != nil {
+				s.cancel(err)
```

Comment:
Why could this cancel prevent callback from being lost?

Reply:
In addition, when WriteEvents returns error, the external caller should also close dml_sink?

Reply:
Here we cancel the sink backend. Then when checking whether the table sink has been closed or not, we can know the sink backend is dead, and then we don't need to be blocked on table sink callbacks.